File: apps/desktop/shell/boot.js
/*
 * The boot page's script. Plain ES2022, no build step, no framework, no import: this file
 * has to run when everything else has failed, so it depends on nothing that could fail.
 *
 * It reads one shape — the `BootState` `boot.rs` serialises — and paints it. The state
 * arrives twice over, on purpose: once by asking (`boot_state`), because the page can
 * load after the answer is already known, and then on every change by listening
 * (`boot-state`), because the answer arrives asynchronously and polling for it would be
 * a timer nobody can see. Neither path is a fallback for the other; both always run.
 *
 * Navigation away from this page is done in Rust, not here. A page that navigated itself
 * would have to trust a URL that came over IPC; the Rust side already has the URL, so it
 * is the one that moves the webview.
 */

const api = window.__TAURI__;

const headline = document.getElementById('headline');
const detail = document.getElementById('detail');
const requirement = document.getElementById('requirement');
const doctorSection = document.getElementById('doctor-section');
const doctor = document.getElementById('doctor');
const download = document.getElementById('download');
const shellVersion = document.getElementById('shell-version');
const root = document.querySelector('.boot');

/** Phases that mean "still working"; anything else is a resting state. */
const WAITING = new Set(['checking-node', 'starting-server']);

const HEADLINES = {
  'checking-node': 'looking for Node…',
  'starting-server': 'starting the local server…',
  ready: 'canvas ready',
  'node-missing': 'Node is not installed, or not on PATH',
  'node-too-old': 'this machine has an older Node than Nazar needs',
  'node-unreadable': 'Node answered, but not with a version',
  'server-failed': 'the local server did not start',
};

function paint(state) {
  if (!state || typeof state.phase !== 'string') return;

  headline.textContent = HEADLINES[state.phase] ?? state.phase;
  root.classList.toggle('boot--waiting', WAITING.has(state.phase));
  root.classList.toggle('boot--failed', state.phase.startsWith('node-') || state.phase === 'server-failed');

  const message = typeof state.detail === 'string' ? state.detail.trim() : '';
  detail.textContent = message;
  detail.hidden = message.length === 0;

  requirement.hidden = !state.phase.startsWith('node-');

  const report = typeof state.doctor === 'string' ? state.doctor.trim() : '';
  doctor.textContent = report;
  doctorSection.hidden = report.length === 0;

  if (typeof state.version === 'string') {
    shellVersion.textContent = `Nazar desktop ${state.version}`;
  }
}

// The install link opens in the machine's browser rather than in this window: a webview
// that navigates to nodejs.org is a webview with no way back to this page. The URL comes
// from Rust so that the one written in `node.rs` is the one shown here.
download.addEventListener('click', (event) => {
  event.preventDefault();
  api?.core?.invoke('open_download_page').catch(() => {
    // Nothing to fall back to but the text of the link, which is already on screen.
  });
});

async function main() {
  if (api?.event?.listen) {
    await api.event.listen('boot-state', (event) => paint(event.payload));
  }
  if (api?.core?.invoke) {
    paint(await api.core.invoke('boot_state'));
  } else {
    // Opened outside the shell — a developer with the file in a browser. Say so rather
    // than sitting on "starting…" forever.
    headline.textContent = 'this page is the desktop shell’s, and needs the desktop shell';
    root.classList.remove('boot--waiting');
  }
}

void main();

File: apps/desktop/src/boot.rs
//! What the window says before the canvas is there, and when it never will be.
//!
//! One value, one event. The boot thread replaces [`BootState`] as it goes; every
//! replacement is emitted as `boot-state`, and the page can also ask for the current one
//! with the `boot_state` command because it may have loaded after the answer was known.
//!
//! The phases are strings rather than a Rust enum with a serde tag because the page keys a
//! lookup table on them, and a string that arrives without a translation shows itself
//! instead of showing nothing.

use serde::Serialize;

/// Everything the boot page paints.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BootState {
    /// `checking-node`, `starting-server`, `ready`, `node-missing`, `node-too-old`,
    /// `node-unreadable` or `server-failed`.
    pub phase: &'static str,
    /// A sentence under the headline. Empty when there is nothing to add.
    pub detail: String,
    /// `nazar doctor` output, when it could be produced. Empty otherwise.
    pub doctor: String,
    /// The shell's version, shown in the footer.
    pub version: String,
    /// The canvas URL, once there is one.
    pub url: String,
}

/// Phase constants, so a typo is a compile error rather than a blank headline.
pub mod phase {
    /// Running `node --version`.
    pub const CHECKING_NODE: &str = "checking-node";
    /// Waiting for the child server to bind and answer.
    pub const STARTING_SERVER: &str = "starting-server";
    /// The canvas is up; the webview is on its way there.
    pub const READY: &str = "ready";
    /// `node` could not be run at all.
    pub const NODE_MISSING: &str = "node-missing";
    /// `node` ran and is older than the packaged server needs.
    pub const NODE_TOO_OLD: &str = "node-too-old";
    /// `node` ran but printed no version.
    pub const NODE_UNREADABLE: &str = "node-unreadable";
    /// Node was fine and the server still did not come up.
    pub const SERVER_FAILED: &str = "server-failed";
}

impl BootState {
    /// The state a window opens in.
    #[must_use]
    pub fn starting(version: &str) -> Self {
        BootState {
            phase: phase::CHECKING_NODE,
            detail: String::new(),
            doctor: String::new(),
            version: version.to_owned(),
            url: String::new(),
        }
    }

    /// The same state in a later phase, with a new sentence.
    #[must_use]
    pub fn moved_to(&self, phase: &'static str, detail: impl Into<String>) -> Self {
        BootState {
            phase,
            detail: detail.into(),
            doctor: self.doctor.clone(),
            version: self.version.clone(),
            url: self.url.clone(),
        }
    }

    /// The same state with `nazar doctor` attached.
    #[must_use]
    pub fn with_doctor(mut self, report: Option<String>) -> Self {
        self.doctor = report.unwrap_or_default();
        self
    }

    /// The same state with the canvas URL attached.
    #[must_use]
    pub fn with_url(mut self, url: &str) -> Self {
        self.url = url.to_owned();
        self
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_phase_change_keeps_what_was_already_known() {
        let start = BootState::starting("0.1.0")
            .with_doctor(Some("two sessions".to_owned()))
            .with_url("http://127.0.0.1:4676");
        let later = start.moved_to(phase::SERVER_FAILED, "it did not bind");

        assert_eq!(later.phase, phase::SERVER_FAILED);
        assert_eq!(later.detail, "it did not bind");
        assert_eq!(
            later.doctor, "two sessions",
            "the report is not lost by a phase change"
        );
        assert_eq!(later.version, "0.1.0");
        assert_eq!(later.url, "http://127.0.0.1:4676");
    }

    #[test]
    fn the_page_and_the_shell_agree_on_every_phase_name() {
        // The page keys its headline table on these. A phase the page has never heard of
        // shows its own name, which is survivable — but a phase renamed on one side only
        // is a regression nobody would notice, so it is checked here.
        let script = std::fs::read_to_string(
            std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
                .join("shell")
                .join("boot.js"),
        )
        .expect("the boot page sits beside this crate");

        for name in [
            phase::CHECKING_NODE,
            phase::STARTING_SERVER,
            phase::READY,
            phase::NODE_MISSING,
            phase::NODE_TOO_OLD,
            phase::NODE_UNREADABLE,
            phase::SERVER_FAILED,
        ] {
            assert!(
                script.contains(name),
                "boot.js has no headline for the {name} phase"
            );
        }
    }

    #[test]
    fn the_serialised_shape_is_the_one_the_page_reads() {
        let json = serde_json::to_string(&BootState::starting("0.1.0")).expect("serialises");
        for key in ["phase", "detail", "doctor", "version", "url"] {
            assert!(
                json.contains(&format!("\"{key}\"")),
                "{key} is missing from {json}"
            );
        }
    }
}
